pygmt/src/x2sys_cross.py

```diff
@@ -183,68 +182,71 @@ def x2sys_cross(tracks=None, outfile=None, **kwargs):
+    for track in tracks:
+        match data_kind(track):
+            case "file":
                 file_contexts.append(contextlib.nullcontext(track))
-            elif kind == "matrix":
+            case "matrix":
                 # find suffix (-E) of trackfiles used (e.g. xyz, csv, etc) from
                 # $X2SYS_HOME/TAGNAME/TAGNAME.tag file
-                lastline = (
-                    Path(os.environ["X2SYS_HOME"], kwargs["T"], f"{kwargs['T']}.tag")
-                    .read_text(encoding="utf8")
-                    .strip()
-                    .split("\n")[-1]
-                )  # e.g. "-Dxyz -Etsv -I1/1"
+                tagfile = Path(
+                    os.environ["X2SYS_HOME"], kwargs["T"], f"{kwargs['T']}.tag"
+                )
+                # Last line is like "-Dxyz -Etsv -I1/1"
+                lastline = tagfile.read_text().splitlines()[-1]
                 for item in sorted(lastline.split()):  # sort list alphabetically
                     if item.startswith(("-E", "-D")):  # prefer -Etsv over -Dxyz
                         suffix = item[2:]  # e.g. tsv (1st choice) or xyz (2nd choice)
 
                 # Save pandas.DataFrame track data to temporary file
                 file_contexts.append(tempfile_from_dftrack(track=track, suffix=suffix))
-            else:
+            case _:
                 raise GMTInvalidInput(f"Unrecognized data type: {type(track)}")
 
-        with GMTTempFile(suffix=".txt") as tmpfile:
+    with Session() as lib:
+        with lib.virtualfile_out(kind="dataset", fname=outfile) as vouttbl:
             with contextlib.ExitStack() as stack:
                 fnames = [stack.enter_context(c) for c in file_contexts]
-                if outfile is None:
-                    outfile = tmpfile.name
                 lib.call_module(
                     module="x2sys_cross",
-                    args=build_arg_list(kwargs, infile=fnames, outfile=outfile),
-                )
-
-            # Read temporary csv output to a pandas table
-            if outfile == tmpfile.name:  # if outfile isn't set, return pd.DataFrame
-                # Read the tab-separated ASCII table
-                date_format_kwarg = (
-                    {"date_format": "ISO8601"}
-                    if Version(pd.__version__) >= Version("2.0.0")
-                    else {}
+                    args=build_arg_list(kwargs, infile=fnames, outfile=vouttbl),
                 )
-                table = pd.read_csv(
-                    tmpfile.name,
-                    sep="\t",
-                    header=2,  # Column names are on 2nd row
-                    comment=">",  # Skip the 3rd row with a ">"
-                    parse_dates=[2, 3],  # Datetimes on 3rd and 4th column
-                    **date_format_kwarg,  # Parse dates in ISO8601 format on pandas>=2
+                result = lib.virtualfile_to_dataset(
+                    vfname=vouttbl, output_type=output_type, header=2
                 )
```

Comment:
Note that `x2sys_cross` can output multi-segment files (each segment is separated by `IO_SEGMENT_MARKER` which is `>` by default, see https://docs.generic-mapping-tools.org/6.5/reference/file-formats.html#optional-segment-header-records). If I'm not mistaken, the current `virtualfile_to_dataset` method does not implement multi-segment file handling yet? To be fair though, the current implementation in `x2sys_cross` simply merges all segments into one, since we skip rows starting with `>`, but we need to check that `virtualfile_to_dataset` will return all segments in a multi-segment file instead of just the first one.

Reply:
> If I'm not mistaken, the current `virtualfile_to_dataset` method does not implement multi-segment file handling yet? To be fair though, the current implementation in `x2sys_cross` simply merges all segments into one, since we skip rows starting with `>`, but we need to check that `virtualfile_to_dataset` will return all segments in a multi-segment file instead of just the first one.

Yes. The main problem is that, as far as I know, there is no equivalent way to represent a multi-segment file in pandas. The multi-segment support was also mentioned in https://github.com/GenericMappingTools/pygmt/pull/2729#discussion_r1378375843. 

If we can have a general way to represent multi-segment in pandas, then it should be straightforward to output multi-segments from `_GMT_DATASET` to the desired data structure.

Reply:
> To be fair though, the current implementation in `x2sys_cross` simply merges all segments into one, since we skip rows starting with `>`, but we need to check that `virtualfile_to_dataset` will return all segments in a multi-segment file instead of just the first one.

It's already tested in the `_GMT_DATASET` docstrings https://github.com/GenericMappingTools/pygmt/blob/466c8b6b45d7d6aa202077d91764bd2ab8ed7324/pygmt/datatypes/dataset.py#L215. 

For `x2sys_cross`, we also test the shape of the output pd.DataFrame.